PR: https://github.com/odoo/odoo/pull/218116

From: 77a74d4eea3df83ab53f2669b44162ec8a885a53
From: unka-odoo
Date: 2025-09-03 19:31:37

Breaking data model changes scores: del:2 + add:0, change matches:
-    account_sale_receipt_tax_id = fields.Many2one('account.tax', string="Default Sale Receipt Tax", check_company=True)
-    account_purchase_receipt_tax_id = fields.Many2one('account.tax', string="Default Purchase Receipt Tax", check_company=True)

Total Changes: 41

[IMP] account,l10n_be:  Default fiscal position on purchase receipts

* Before this commit:
  Localizations can set a Default Taxes that would override the current Default
  Taxes when using a receipt. And for belgium 0% S tax is set as a Default Tax
  for receipt.

* After this commit:
  Instead of allowing localizations to set Default Taxes for receipt, we are now
  allowing localizations to set Default Fiscal Positions that would override the
  current Default Fiscal Position when using a Receipt.

  For belgium new fiscal position 'Non Deductible' have been added in the data
  and Non Deductible version of domestic taxes have also been added with fiscal
  position set as a newly added 'Non Deductible' fiscal position (like:
  21% M ND, 12% M ND, 6% M ND, 12% S ND, 6% S ND, 12% IG ND, 6% IG ND).

closes odoo/odoo#218116

Task: 4907662
Related: odoo/upgrade#8036
Signed-off-by: Habib Ayob (ayh) <ayh@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account/models/account_move.py
+++ b/addons/account/models/account_move.py
@@ -997,9 +997,15 @@ class AccountMove(models.Model):
             else:
                 move.partner_shipping_id = False
 
-    @api.depends('partner_id', 'partner_shipping_id', 'company_id')
+    @api.depends('partner_id', 'partner_shipping_id', 'company_id', 'move_type')
     def _compute_fiscal_position_id(self):
         for move in self:
+            receipt_fiscal_position = {
+                'in_receipt': move.company_id.account_purchase_receipt_fiscal_position_id,
+            }.get(move.move_type)
+            if receipt_fiscal_position:
+                move.fiscal_position_id = receipt_fiscal_position
+                continue
             delivery_partner = self.env['res.partner'].browse(
                 move.partner_shipping_id.id
                 or move.partner_id.address_get(['delivery'])['delivery']

--- a/addons/account/models/account_move_line.py
+++ b/addons/account/models/account_move_line.py
@@ -961,27 +961,21 @@ class AccountMoveLine(models.Model):
         self.ensure_one()
 
         company_domain = self.env['account.tax']._check_company_domain(self.move_id.company_id)
-        tax_ids = self.env['account.tax']
+        if self.move_id.is_sale_document(include_receipts=True):
+            # Out invoice.
+            filtered_taxes_id = self.product_id.taxes_id.filtered_domain(company_domain)
+            tax_ids = filtered_taxes_id or self.account_id.tax_ids.filtered(lambda tax: tax.type_tax_use == 'sale')
 
-        if self.move_id.is_receipt():
-            tax_ids = self.company_id.account_purchase_receipt_tax_id if self.move_id.move_type == 'in_receipt' else self.company_id.account_sale_receipt_tax_id
+        elif self.move_id.is_purchase_document(include_receipts=True):
+            # In invoice.
+            filtered_supplier_taxes_id = self.product_id.supplier_taxes_id.filtered_domain(company_domain)
+            tax_ids = filtered_supplier_taxes_id or self.account_id.tax_ids.filtered(lambda tax: tax.type_tax_use == 'purchase')
 
-        if not tax_ids:
-            if self.move_id.is_sale_document(include_receipts=True):
-                # Out invoice.
-                filtered_taxes_id = self.product_id.taxes_id.filtered_domain(company_domain)
-                tax_ids = filtered_taxes_id or self.account_id.tax_ids.filtered(lambda tax: tax.type_tax_use == 'sale')
+        elif self.env.context.get('account_default_taxes'):
+            tax_ids = self.account_id.tax_ids
 
-            elif self.move_id.is_purchase_document(include_receipts=True):
-                # In invoice.
-                filtered_supplier_taxes_id = self.product_id.supplier_taxes_id.filtered_domain(company_domain)
-                tax_ids = filtered_supplier_taxes_id or self.account_id.tax_ids.filtered(lambda tax: tax.type_tax_use == 'purchase')
-
-            elif self.env.context.get('account_default_taxes'):
-                tax_ids = self.account_id.tax_ids
-
-            else:
-                tax_ids = False if self.env.context.get('skip_computed_taxes') or self.move_id.is_entry() else self.account_id.tax_ids
+        else:
+            tax_ids = False if self.env.context.get('skip_computed_taxes') or self.move_id.is_entry() else self.account_id.tax_ids
 
         if self.company_id and tax_ids:
             tax_ids = tax_ids._filter_taxes_by_company(self.company_id)

--- a/addons/account/models/company.py
+++ b/addons/account/models/company.py
@@ -122,8 +122,7 @@ class ResCompany(models.Model):
     transfer_account_code_prefix = fields.Char(string='Prefix of the transfer accounts')
     account_sale_tax_id = fields.Many2one('account.tax', string="Default Sale Tax", check_company=True)
     account_purchase_tax_id = fields.Many2one('account.tax', string="Default Purchase Tax", check_company=True)
-    account_sale_receipt_tax_id = fields.Many2one('account.tax', string="Default Sale Receipt Tax", check_company=True)
-    account_purchase_receipt_tax_id = fields.Many2one('account.tax', string="Default Purchase Receipt Tax", check_company=True)
+    account_purchase_receipt_fiscal_position_id = fields.Many2one('account.fiscal.position', string="Default Purchase Receipt Fiscal Position", check_company=True)
     tax_calculation_rounding_method = fields.Selection([
         ('round_globally', 'Round per Tax'),
         ('round_per_line', 'Round per Line'),
